presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import static com.facebook.presto.hive.util.TempFileWriter.appendPositionTo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MergingPageIterator
+        extends AbstractIterator<Page>
+{
+    private final Queue<PagePositionIterator> queue = new PriorityQueue<>();
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final PageBuilder pageBuilder;
+
+    public MergingPageIterator(
+            Collection<Iterator<Page>> iterators,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders)
+    {
+        requireNonNull(sortFields, "sortFields is null");
+        requireNonNull(sortOrders, "sortOrders is null");
+        checkArgument(sortFields.size() == sortOrders.size(), "sortFields and sortOrders size must match");
+
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(sortFields);
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.pageBuilder = new PageBuilder(types);
+
+        iterators.stream()
+                .map(PagePositionIterator::new)
+                .filter(PagePositionIterator::hasNext)
+                .forEach(queue::add);
+    }
+
+    @Override
+    protected Page computeNext()
+    {
+        while (!pageBuilder.isFull()) {
+            PagePositionIterator iterator = queue.poll();
+            if (iterator == null) {
+                break;
+            }
+
+            iterator.appendNextTo(pageBuilder);
+
+            if (iterator.hasNext()) {
+                queue.add(iterator);
+            }
+        }
+
+        if (pageBuilder.isEmpty()) {
+            return endOfData();
+        }
+
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
+    }
+
+    @SuppressWarnings("ComparableImplementedButEqualsNotOverridden")
+    private class PagePositionIterator
```

Comment:
it would be better to use compositions of iterators, e.g:
```
  Iterator<Page>
  -> Iterator<PageWithPosition>
  -> merge -> Iterator<PageWithPosition>
  -> build page -> Iterator<Page>
```

Reply:
You're right, I switched this to use `Iterator<PagePosition>` and the code is simpler. The merge algorithm is now basically
```java
mergeSorted(... concat(transform(pages, PagePositionIterator::new))))